Strict CSP article for web.dev (#4862)
GoogleChrome/web.dev: src/site/content/en/blog/strict-csp/index.md

---
title: Mitigate cross-site scripting (XSS) with a strict Content Security Policy (CSP)
subhead: How to deploy a CSP based on script nonces or hashes as a defense-in-depth against cross-site scripting.
authors:
  - lwe
date: 2021-03-15
# updated: 
hero: image/3lmWcR1VGYVMicNlBh4aZWBTcSg1/mhE0NYvP3JFyvNyiQ1dj.jpg
alt: A screenshot of JavaScript code setting a strict Content Security Policy.
tags:
  - blog
  - security

>   - csp

Reviewer: ```suggestion
```

---

## Why should you deploy a strict Content Security Policy (CSP)?

[Cross-site scripting (XSS)](https://www.google.com/about/appsecurity/learning/xss/)—the ability
to inject malicious scripts into a web application—has been one of the biggest web security vulnerabilities for over a decade.

> Content Security Policy (CSP) is an added layer of security that helps to mitigate XSS. Configuring a CSP involves adding the Content-Security-Policy HTTP header to a web page

Reviewer: ```suggestion
[Content Security Policy (CSP)](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) is an added layer of security that helps to mitigate XSS. Configuring a CSP involves adding the Content-Security-Policy HTTP header to a web page
```

and giving it values to control what resources the user agent is allowed to load for that page.
This article explains how to use a CSP based on nonces or hashes to mitigate XSS instead of
the commonly used host allowlist based CSPs which often leave the page exposed to XSS as they can be [bypassed in most configurations](https://research.google.com/pubs/pub45542.html).

{% Aside 'key-term' %}
A _nonce_ is a random number used only once that can be used to mark a `<script>` tag as
trusted.
{% endAside %}

{% Aside 'key-term' %}
A hash function is a mathematical function that converts an input value into a compressed
numerical value–a hash. A _hash_ (e.g. sha256) can be used to mark an inline `<script>` tag as
trusted.
{% endAside %}

> A [Content Security Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) based on

Reviewer: ```suggestion
A Content Security Policy based on
```

---
title: Mitigate cross-site scripting (XSS) with a strict Content Security Policy (CSP)
subhead: How to deploy a CSP based on script nonces or hashes as a defense-in-depth against cross-site scripting.

> authors:

Reviewer: ```suggestion
description: |
    Learn how to deploy a CSP based on script nonces or hashes as a defense-in-depth 
    against cross-site scripting.
authors:
```

party JavaScript libraries and widgets.
- Not based on URL allowlists and therefore doesn't suffer from [common CSP
bypasses](https://speakerdeck.com/lweichselbaum/csp-is-dead-long-live-strict-csp-deepsec-2016?slide=15). 
- Blocks untrusted inline scripts like inline event handlers or `javascript:` URIs.
- Restricts `object-src` to disable dangerous plugins such as Flash.
- Restricts `base-uri` to block the injection of `<base>` tags. This prevents attackers from
changing the locations of scripts loaded from relative URLs.

{% Aside %}
Another advantage of a strict CSP is that the CSP always has the same structure and doesn't
need to be customized for your application.
{% endAside %}

## Adopting a strict CSP

To adopt a strict CSP, you need to:
1. Decide if your application should set a nonce- or hash-based CSP.
1. Copy the CSP from the [What is a strict Content Security Policy](#what-is-a-strict-content-security-policy) section and set it as a response header across
your application. 
1. Refactor HTML templates and client-side code to remove patterns that are incompatible with
CSP.
1. Add fallbacks to support Safari and older browsers.
1. Deploy your CSP.

> You can use [Lighthouse](https://github.com/GoogleChrome/lighthouse) throughout this process to check whether your site has a CSP,

Reviewer: ```suggestion
You can use [Lighthouse](https://developers.google.com/web/tools/lighthouse) throughout this process to check whether your site has a CSP,
```

party JavaScript libraries and widgets.
- Not based on URL allowlists and therefore doesn't suffer from [common CSP
bypasses](https://speakerdeck.com/lweichselbaum/csp-is-dead-long-live-strict-csp-deepsec-2016?slide=15). 
- Blocks untrusted inline scripts like inline event handlers or `javascript:` URIs.
- Restricts `object-src` to disable dangerous plugins such as Flash.
- Restricts `base-uri` to block the injection of `<base>` tags. This prevents attackers from
changing the locations of scripts loaded from relative URLs.

{% Aside %}
Another advantage of a strict CSP is that the CSP always has the same structure and doesn't
need to be customized for your application.
{% endAside %}

## Adopting a strict CSP

To adopt a strict CSP, you need to:
1. Decide if your application should set a nonce- or hash-based CSP.
1. Copy the CSP from the [What is a strict Content Security Policy](#what-is-a-strict-content-security-policy) section and set it as a response header across
your application. 
1. Refactor HTML templates and client-side code to remove patterns that are incompatible with
CSP.
1. Add fallbacks to support Safari and older browsers.
1. Deploy your CSP.

> You can use [Lighthouse](https://developers.google.com/web/tools/lighthouse) **Best Practices** audit throughout this process to check whether your site has a CSP, and whether it's strict enough to be effective against XSS. 

Reviewer: ```suggestion
You can use [Lighthouse](https://developers.google.com/web/tools/lighthouse) (v7.3.0 and above) **Best Practices** audit throughout this process to check whether your site has a CSP, and whether it's strict enough to be effective against XSS. 
```

party JavaScript libraries and widgets.
- Not based on URL allowlists and therefore doesn't suffer from [common CSP
bypasses](https://speakerdeck.com/lweichselbaum/csp-is-dead-long-live-strict-csp-deepsec-2016?slide=15). 
- Blocks untrusted inline scripts like inline event handlers or `javascript:` URIs.
- Restricts `object-src` to disable dangerous plugins such as Flash.
- Restricts `base-uri` to block the injection of `<base>` tags. This prevents attackers from
changing the locations of scripts loaded from relative URLs.

{% Aside %}
Another advantage of a strict CSP is that the CSP always has the same structure and doesn't
need to be customized for your application.
{% endAside %}

## Adopting a strict CSP

To adopt a strict CSP, you need to:
1. Decide if your application should set a nonce- or hash-based CSP.
1. Copy the CSP from the [What is a strict Content Security Policy](#what-is-a-strict-content-security-policy) section and set it as a response header across
your application. 
1. Refactor HTML templates and client-side code to remove patterns that are incompatible with
CSP.
1. Add fallbacks to support Safari and older browsers.
1. Deploy your CSP.

> You can use [Lighthouse](https://developers.google.com/web/tools/lighthouse) throughout this process to check whether your site has a CSP,
> and whether it's strict enough to be effective against XSS.
> 
> {% Img src="image/vgdbNJBYHma2o62ZqYmcnkq3j0o1/DAt65SsdOKxEHmtNFEKX.png", alt="Lighthouse report warning that no CSP is found in enforcement mode.", width="800", height="174", class="w-screenshot" %}

Reviewer: ```suggestion
You can use [Lighthouse](https://developers.google.com/web/tools/lighthouse) **Best Practices** audit throughout this process to check whether your site has a CSP, and whether it's strict enough to be effective against XSS. 

{% Img src="image/vgdbNJBYHma2o62ZqYmcnkq3j0o1/DAt65SsdOKxEHmtNFEKX.png", alt="Lighthouse report warning that no CSP is found in enforcement mode.", width="800", height="174", class="w-screenshot" %}
```

(nonce) for every response.</td>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>Hash-based CSP</th>
        <td>For HTML pages served statically or those that need to be cached. For example,
single-page web applications built with frameworks such as Angular, React or others, that are
statically served without server-side rendering.</td>
      </tr>
    </tbody>
  </table>
</div>

### Step 2: Set a strict CSP and prepare your scripts

When setting a CSP, you have a few options:
- Report-only mode (`Content-Security-Policy-Report-Only`) or enforcement mode
  (`Content-Security-Policy`). In report-only, the CSP won't block resources
  yet—nothing will break—but you'll be able to see errors and receive reports
  for what would have been blocked. Locally, when you're in the process of
  setting a CSP, this doesn't really matter, because both modes will show you
  the errors in the browser console. If anything, enforcement mode will make it
  even easier for you to see blocked resources and tweak your CSP, since your
  page will look broken. Report-only mode becomes most useful later in the

>   process (see [Step 5](#step-5:-deploy-your-csp). 

Reviewer: ```suggestion
  process (see [Step 5](#step-5:-deploy-your-csp)). 
```
